src/main/java/com/google/api/codegen/gapic/GapicProviderFactory.java

```
@@ -0,0 +1,202 @@
+import com.google.api.codegen.ApiConfig;
+import com.google.api.codegen.InterfaceListView;
+import com.google.api.codegen.InterfaceView;
+import com.google.api.codegen.ProtoFileView;
+import com.google.api.codegen.clientconfig.ClientConfigGapicContext;
+import com.google.api.codegen.clientconfig.ClientConfigSnippetSetRunner;
+import com.google.api.codegen.csharp.CSharpGapicContext;
+import com.google.api.codegen.csharp.CSharpSnippetSetRunner;
+import com.google.api.codegen.go.GoGapicContext;
+import com.google.api.codegen.go.GoSnippetSetRunner;
+import com.google.api.codegen.java.JavaGapicContext;
+import com.google.api.codegen.java.JavaIterableSnippetSetRunner;
+import com.google.api.codegen.java.JavaSnippetSetRunner;
+import com.google.api.codegen.nodejs.NodeJSGapicContext;
+import com.google.api.codegen.nodejs.NodeJSSnippetSetRunner;
+import com.google.api.codegen.php.PhpGapicContext;
+import com.google.api.codegen.php.PhpSnippetSetRunner;
+import com.google.api.codegen.py.PythonGapicContext;
+import com.google.api.codegen.py.PythonInterfaceInitializer;
+import com.google.api.codegen.py.PythonProtoFileInitializer;
+import com.google.api.codegen.py.PythonSnippetSetRunner;
+import com.google.api.codegen.ruby.RubyGapicContext;
+import com.google.api.codegen.ruby.RubySnippetSetRunner;
+import com.google.api.tools.framework.model.Interface;
+import com.google.api.tools.framework.model.Model;
+import com.google.api.tools.framework.model.ProtoFile;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.apache.commons.lang3.NotImplementedException;
+
+/**
+ * GapicProviderFactory creates GapicProvider instances based on an id.
+ */
+public class GapicProviderFactory {
+
+  /**
+   * Create the GapicProviders based on the given id
+   */
+  public static List<GapicProvider<? extends Object>> defaultCreate(Model model,
+      ApiConfig apiConfig, String id) {
+
+    // Please keep the following IDs in alphabetical order
+
+    if (id.equals("client_config")) {
```

I don't think that really buys us anything; it creates an extra layer of indirection for something that will fail fast if it's wrong (the system won't even start).